[sully] HTML 코드 중복 제거 / 회원 가입 / 회원 목록 조회 / 회원 프로필 정보 조회 / 게시글 작성 / 게시글 목록 조회 / 게시글 상세 조회 / 회원 정보 수정 화면

devbattery opened the pull request:
# 학습 과정

- 구글링만으로 구현해보려다가 이건 도저히 안 될 것 같아서 영한님의 mvc 강의를 듣는 데 이틀을 썼습니다.
- 확실히 듣기 전과 비교하면 전체적인 감은 잡힌 것 같습니다. 그래도 구현은 완전히 다른 문제였습니다.
- 머스타치를 사용하려 했지만 이러다 구현 시작조차 하지 못할 것 같아서 강의에서 활용했던 타임리프를 사용했습니다.

# 주요 고민과 해결
- 스프링에 대한 지식이 아예 없는 상태에서 구현을 시작했다 보니 Mapping의 개념조차 이해하기 힘들었습니다.
  - 직접 html 파일과 컨트롤러를 작은 부분부터 연결하면서 연습했습니다.
  - 이러한 과정에서 GetMapping, PostMapping과 html 연결이 익숙해졌습니다.
- 영한님의 강의에서 짜여진 코드를 보고도 구현에 옮기기 쉽지 않았습니다.
  - User와 Article의 코드로 각각 직접 바꿔보면서 각 메서드와 클래스들의 관계에 대해 감을 잡을 수 있었습니다.
  - 아직 Form과 Service의 사용은 특히나 더 미숙하지만, 그냥 이런 식으로 쓰인다는 것 자체는 이해할 수 있었습니다.
- 리다이렉트 개념이 HTTP 공부할 때도 그렇고 그저 난해하기만 했습니다.
  - 단순하게 접근하여 회원가입, 게시물 작성으로 경험해 보니 신기하기도 하고 익숙해질 수 있었습니다.
- 타임리프의 문법을 처음 봤을 때, 뭐가 무엇인지 감이 하나도 잡히질 않았습니다.
  - 그걸 가지고 응용까지 해야 하니 더 멘붕이었습니다.
  - 이 부분도 마찬가지로 강의만 따라치는 것이 아니라  여러 상황을 접하면서  직접 프로젝트에 적용시켜 보니
    - 어떤 변수가 넘어와도 개별이나 each를 써서 여러 변수를 관리하는 데 어려움이 줄어들었습니다.
- 동료들의 도움을 많이 받아 결국 2단계까지 구현에 성공했습니다.
  - 도움을 받지 않고 저 혼자만 끙끙댔다면 분명 아직 1단계도 성공하지 못 했을 거라 생각합니다.
  - 모르는 부분은 확실하게 도움을 받고 머리에 새기는 것이 기억에도 더 오래 남고 좋은 것 같습니다.
  - 추가 미션은 시간 부족으로 회원정보 수정 화면으로 이동하는 것까지만 구현하였습니다.
- `Map<String, User>` vs `Map<Long, User>`
  - MemoryArticleRepository 클래스에서는 글 목록의 index를 카운트해줘야 하므로 Long형으로 저항 없이 썼지만
    - MemoryUserRepository 클래스에서는 user의 index가 필요 없으므로 String형으로 사용했습니다.
  - 이 주제에 대하여 동료들과 토론을 했지만 결론은 어떤 상황에서도 Long형으로 정의하는 것이 맞다는 것이었습니다.
    - _**kyu께서는 이에 대해 어떻게 생각하시는지 여쭤보고 싶습니다. 어떤 상황에서든 index를 추가하여 Long형으로 정의하는 것이 맞을까요?**_
  - 이 미션에서는 아직 동시성 문제가 발생할 일이 없으므로 `HashMap<>()`을 사용했습니다.

# 부족한 점

- MVC 패턴에 대해 아직 정확히 파악하고 있지 못 하고 있는 것 같습니다.
  - 영한님 강의의 샘플을 보고 흉내내며 구현하긴 했지만 아예 안 보고 처음부터 구현하라고 하면 힘들 것 같습니다.
  - 저도 "선 이론 후 실습"보다는 "선 실습 후 이론"이 체화가 훨씬 더 잘 되기 때문에 이제부터 이론을 익혀가려 합니다.
- 입력 예외 처리를 거의 하지 못했습니다.
  - 회원가입에서 `userId`를 입력할 때도 같은 아이디를 입력하면 다시 입력하라는 폼이 뜨는 것이 아니라 `Whitelabel Error Page`로 이동합니다.
    - 이유는 UserService에서 `userId`가 이미 존재할 경우 `throw new IllegalStateException()`을 했기 때문입니다.
    - 이렇게 하지 말고 직접 폼에서 다시 입력을 받도록 하고 싶었는데 현재 실력으로는 많이 부족하다고 생각하여 하지 못했습니다.
  - 다른 동료분들이 예외 처리를 한 것을 보면 springboot의 validation 기능을 이용하여 아이디, 비밀번호, 이름, 이메일 형식에 대한 예외처리도 하셨던데 나중에 예외 처리를 할 기회가 있으면 이것을 사용해 볼 계획입니다.

# 구현 결과 GIF

## 회원 가입 및 회원 목록 조회 기능

![회원가입 후 프로필 이동](https://user-images.githubusercontent.com/62871026/229013575-f027831d-c60e-4259-a2d0-5be74b2422f5.gif)

- 입력 정보를 입력하고 회원가입을 할 수 있습니다.
- 회원가입이 성공적으로 이루어지면 그 정보를 가지고 profile로 이동합니다.

<br>
<br>

![프로필에서 리스트 이동](https://user-images.githubusercontent.com/62871026/229044273-ec319b8d-cddc-4163-985d-7427e8777393.gif)


- 오른쪽 위 사람표시를 클릭하면 users로 리다이렉트 되도록 했습니다.
- 프로필에서 리스트를 자유자재로 이동할 수 있습니다.
- 사용자 아이디를 클릭하면 그 아이디의 프로필로 들어갑니다.
  - 구현 2단계에서 Posts의 이름을 클릭하면 프로필로 이동하는 것을 참고했습니다.

<br>
<br>

## 글 쓰기 기능

![글 쓰고 메인페이지에 출력](https://user-images.githubusercontent.com/62871026/229013833-e0dfc6f1-09a5-476d-b7df-c29e49890926.gif)

- 글쓴이, 제목, 내용을 입력하면 index로 리다이렉트 됩니다.
- 리다이렉트 됨과 동시에 글쓴이와 제목 정보가 표시됩니다.

<br>
<br>

![글 쓴 내용을 바탕으로 상세 보기](https://user-images.githubusercontent.com/62871026/229014035-7cbed91c-f080-42a3-acb3-b8a74bde8550.gif)

- 입력했던 것을 바탕으로 글쓴이, 제목, 내용을 표시합니다.
- 타이틀과 글쓴이를 누르면 게시물 상세 보기가 가능합니다.
  - 현재는 로그인 정보가 없으므로 글쓴이를 눌러도 그 게시물로 가게끔 만들었습니다.

<br>
<br>

![게시글 상세 보기](https://user-images.githubusercontent.com/62871026/229055048-f7c0ac87-f187-4119-8599-f2c54cdd5dde.gif)

- 게시글이 추가될 때마다 `articleIndex`가 1씩 늘어납니다.
- 주소창에서 `questions/{articleIndex}`를 확인할 수 있습니다.

<br>
<br>

## (선택미션) 회원정보 수정 기능

![회원 정보 수정 화면](https://user-images.githubusercontent.com/62871026/229043831-9a815847-3c33-42ba-baad-9173ff263f7f.gif)

-  `수정` 버튼을 누르면 `users/update/{userId}` 경로로 넘어가서  `updateForm.html`을 보여주도록 하였습니다.
  - 사용자 리스트의 userId 값에 따라 {userId} 경로가 달라집니다.

Commit:
docs: 학습 계획 작성

Commit:
feat: 마크업 파일 업로드

블로그에 있던 양식을 참고하여 root 폴더에 바로 놨는데 이게 맞는지는 잘 모르겠다

Commit:
fix: 마크업 파일 제거

프로젝트 구조를 공부하고 보니 resources 폴더 내에 static 폴더로 마크업 파일이 존재한다는 것을 깨달았다.

Commit:
feat: 카카오 마크업 파일로 변경

원래 위치한 static 폴더는 미션에서 주어진 디자인이 아니었다.

Commit:
feat: template 파일인 mustache 적용 시도

mustache 플러그인 설치 후 templates 폴더 안에 파일이 위치시키게 했다. 나중을 위한 발판이다.

Commit:
fix: 기존 qna 폴더로 다시 수정

호눅스와의 클래스를 갖고 난 후 제일 쉬운 방법이 기존의 폴더 방식이라는 것을 알게 되어서 기본 코드에 있는 폴더로 다시 수정했다.

kyupid commented:
안녕하세요 처음 뵙겠습니다 sully~ 술리라고 발음하면 되나요?    
질문이 있는데 혹시 이 PR 로 계속 수정한다는 말씀이 어떤 의미일까요? 같은 PR 에서 5주 과정을 진행한다는 말씀이신걸까요?

devbattery commented:
@kyupid 

안녕하세요 kyu! 설리라고 발음하시면 됩니다 :)

아까 호눅스께서 1주마다 하나의 PR로 작성하라 하셔서
1주동안 PR의 comment를 계속 수정한다는 의미로 작성해봤습니다!

원래는 1주마다 step-1,2,3,4 등으로 PR을 나눠 요청을 했어서 그렇습니다 !

Commit:
feat: templates 폴더에 form.html, index.html 복사 후 추가

동적 회원가입 기능 구현을 위해 복사

Commit:
feat: 타임리프 설정 후 폴더 위치 변경

gradle과 각 html 파일에 타임리프 코드를 추가해주고 폴더 위치를 static에서 templates로 옮겼다.

Commit:
feat: 사용자가 전달한 값을 관리하는 User 클래스 생성

각 필드를 생성한 후 setter와 getter 생성

Commit:
fix: userId를 id로 변경

내가 한 것은 index였고, 여기서 해야 하는 것은 아이디 그 고유의 값이므로 String형으로 변경

Commit:
style: id의 변수명 변경

html 파일을 보다 보니 id가 아닌 userId로 작성되어 있는 것을 발견
변수명만 간단하게 변경해 줌

Commit:
feat: 회원가입 버튼을 누르면 form.html로 넘어가게끔 구현

이제 코드 중복을 제거하고 매핑하는 데 문제가 없도록 할 예정입니다.

Commit:
feat: 가장 하단에 위치한 중복 코드 분리

header.html이라는 네이밍

Commit:
feat: 가장 상단에 위치한 중복 코드 분리

header.html이라는 네이밍

Commit:
feat: footer와 header를 불러오도록 구현

th:replace를 이용하여 분리한 코드를 불러옴

Commit:
frag: navbar-collapse 코드 분리 후 적용

index.html에서 user/join으로 수정해줬던 그 코드이다. 모든 코드에서 중복이 되어서 다 수정했다.

Commit:
fix: "."을 제거하였고, footer로 되어 있던 것을 수정

</div>.로 되어 있어서 그 부분을 수정하였고
index.html에서 navbar-collapse.html를 불러와야 할 부분에 footer.html이 불러오도록 짜여져 있었음

Commit:
style: navbar-collapse에서 navbar-collapse1로 변수명 변경

navbar-collapse2를 위해 변경

Commit:
frag: navbar-collapse1 코드 분리 후 적용

navbar-collapse2와 똑같은 방식이다

Commit:
frag: subnav.html 코드 분리 후 적용

Commit:
frag: navbar-fixed-top.html 코드 분리 후 적용

Commit:
feat: 유저 리포지토리 인터페이스 생성

MemoryUserRepository에서 이 인터페이스를 구현

Commit:
feat: save 메서드의 반환값을 Optional<>로 변경

값이 안 들어올 수 있음

Commit:
feat: UserRepository를 구현하는 MemoryUserRepository 클래스 생성 후 save 메서드 구현

Map의 key는 userId, value는 User 객체를 받는 HashMap을 생성
단, 실무에서는 동시성 문제 때문에 ConcurrentHashMap을 사용하지만 db에서 없애줄 예정이므로 HashMap 사용
userMap에 넣어줄 때 이미 key값이 존재한다면 Optional의 empty를 반환하도록 함

Commit:
feat: find 역할의 기능 추가

UserId와 객체를 반환해주는 메서드 두개를 구현했다. 항상 userId는 null인 경우를 생각한다.

kyupid commented on README.md:
```diff
@@ -1,2 +1,19 @@
+    - [ ] `redirect`에 대해 공부하기
+    - [ ] GET 요청과 POST 요청에 대한 것을 실습으로 이해하기
+- [ ] 주어진 마크업 파일을 이용하는 방법 공부하기
+    - [ ] `Mustache`에 대해 공부하기
```
Mustache 나 사용하신 Thymleaf 같은 템플릿 엔진이 어떤 식으로 동작하는지 배워보시면 좋을 것 같아요. 서버 사이드 렌더링 (SSR) 과 클라이언트 사이드 렌더링 (CSR) 가 무엇인지 조사해보시고 비교도 해보시면 좋은 공부가 될 것 같습니다.

devbattery replied:
앞으로 갈 길이 멀지만 타임리프 계속 사용해보면서 어떤 식으로 동작하는지 원리부터 차근차근 배우겠습니다.

<br>

- 클라이언트 사이트 렌더링: 사용자의 요청에 따라 필요한 부분만 응답 받아 렌더링하는 방식
- 서버 사이드 렌더링: 서버로부터 완전하게 만들어진 html 파일을 받아와 페이지 전체를 렌더링 하는 방식

일단은 큰 틀로만 비교해 보았는데, 앞으로 구현해 나가면서 자세히 어떤 것인지 꼭 공부해나가겠습니다 감사합니다!

kyupid commented on README.md:
```diff
@@ -1,2 +1,19 @@
+    - [ ] `Mustache`에 대해 공부하기
+    - [ ] HTML 중복 제거에 대해 생각하기
+- [ ] 어느 정도 익숙해지면 실습에 필요한 강의만 찾아 보기
+- [ ] 이번주부터 리뷰 받는다는 생각을 항상 하면서 커밋과 커밋 메시지 하나하나 신중하게 작성하기
```
네. 나름대로 어떤 단위만큼 한 커밋으로 가져갈지 나름의 전략을 세워보는 것도 좋겠네요.

devbattery replied:
감사합니다. 나름 세분화해서 커밋 했다고 생각했는데 다시 보니 난해한 부분이 많은 것 같습니다.
말씀하신 대로 어떤 단위로 할지 전략을 잘 세워나가야 될 것 같습니다.

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,12 @@
+public class UserController {
+    @GetMapping("/user/join")
+    public String joinForm() {
+        return "user/form";
```
ViewResolver 에 대해 혹시 아시나요? 

절대경로로 리턴하는 것과 상대경로로 리턴하는 것의 차이점은 무엇일까요?

devbattery replied:
ViewResolver에 대해서 많은 영상도 찾아보고 구글링도 해보고 이해하려는 노력을 많이 해봤지만
사실 아직은 이해도가 많이 부족한 것 같습니다.
컨트롤러가 뷰 이름을 이용하여 뷰 객체를 찾아주는 역할을 하는 것이 ViewResolver이고,
컨트롤라가 그 뷰 이름만 지정하고, ViewResolver가 이를 이용하여 화면을 생성한다는 이론적인 지식으로만 알고 있습니다.

이번에 스프링에 대해 처음 공부하다 보니 강의를 다 들으면서 구현을 할 수가 없어서 실습 부분만 찾아 보다 보니
응용과 이해가 떨어지는 부분이 많이 발생한 것 같습니다.

`/user/join`처럼 제가 작성한 것은 상대경로이고, 전체적인 경로를 다 적는 것이 절대 경로인데
상대 경로로 리턴하면 현재 위치의 경로를 신경쓰지 않아도 된다고 생각합니다!

kyupid replied:
`user/form` String 을 반환하면 자동으로 resources 아래에 templates/user/form.html 에 있는 페이지를 보여줍니다. 거기서 ViewResolver 가 하는 역할을 여쭤보고 싶었어요. 절대 경로(`/user/form`)로 적으면 동작을 안할텐데 그 원리를 알면 좋을 것 같아서 언급했어요. 참고로 타임리프를 사용하셨으니 ThymleafViewResolver 가 동작하겠죠?

devbattery replied:
감사합니다 ㅠㅠㅠㅠ
그 질문 던져주신 덕분에 지금 다른 동료분께 코드 리뷰 요청드려서 이 부분 완벽하게 이해 받을 수 있었어요
`return "user/list";` <- 이 반환값을 gradle에 있는 타임리프가
` return "templates/" + "user/list" + ".html"` <- 이렇게 바뀌준다고 하더라구요
그때 ThymleafViewResolver가 동작한다고 이해하겠습니다!

kyupid replied:
(수정)

user/form String 을 반환하면 자동으로 resources 아래에 templates/user/form.html 에 있는 페이지를 보여줍니다. 거기서 ViewResolver 가 하는 역할을 여쭤보고 싶었어요.

예전에 타임리프 쓰면서 개발할 때 절대경로로 모두 리턴했었는데 배포하고나면 에러가 나서 모두 상대경로로 변경해주었더니 잘 동작했었거든요. (https://okky.kr/questions/1146193)
그래서 user/form 이랑 /user/form 이 기능적으로 다르게 동작하는 줄 알고 그 차이점도 언급하고 싶었는데 저도 기억이 잘 ㅎㅎ

이건 현재는 다른 부분에 비해서 자세히 알필요는 없으니 지금 알고있는 것으로 충분한 것 같습니다

devbattery replied:
아! 일단 항상 상대 경로로 생각하는 게 좋겠네요
저도 Mapping 할 때는 `/user/xx/` 이런 식이고
리턴할 때는 `user/xx/` 이런 식이어서 헷갈렸는데 나중에 생각나시면 말씀 부탁드려요! 😁

kyupid commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public void setEmail(String email) {
+        this.email = email;
```
User 클래스의 getter 와 setter 들은 왜 선언했고 어떤 역할을 기대하고 만드셨나요?

devbattery replied:
아래의 미션 구현 조건을 보고 User 클래스에 선언된 필드들을 가지고 변수 설정과 받아오는 일들이 필요하다 생각했습니다.
게터는 어쩌다 쓸 일이 있어도, 세터 사용은 최대한 자제하라는 사실을 알고 있어서
이번은 작은 프로젝트라 괜찮겠지만 나중에 프로젝트가 커진다면 어떻게 해야 할지 고민이 됐습니다.

> 사용자가 전달한 값을 User 클래스를 생성해 저장한다. <br>
> 회원가입할 때 전달한 값을 저장할 수 있는 필드를 생성한 후 setter와 getter 메소드를 생성한다.

kyupid replied:
네 생성자도 적절히 사용해서 말씀하신 대로 setter 는 필요할 때만 사용하는 게 좋을 것 같아요.

devbattery replied:
감사합니다!! 지금 회원목록 기능까지 구현 완료했는데
이 코드에서 setter와 dto가 필요한가 아닌가에 대해서까지 수정해서 올려보겠습니다!

kyupid reviewed (commented):
안녕하세요 설리 첫번째 주가 시작되었네요. 화이팅입니닷👍

아직 완료하신 건 아닌 것 같지만 조금 코멘트를 달아놨습니다.

그리고 아래 설명이 잘 이해가 안가요. 무엇이 잘못되어서 id 를 변경했다는 것인지 좀 더 자세히 말씀해주실 수 있으실까요?
```markdown
domain 패키지를 만들어 User 클래스에서 게터와 세터를 생성해주었습니다.
코드 리뷰에서 동료들의 도움으로 잘못된 점을 파악할 수 있었습니다.
id를 private Long id로 설정을 해주었는데 이를 private String userId로 수정하였습니다.
영한님 강의에서 본 것을 그대로 썼던 것이 원인이었고, 이는 그저 목록의 인덱스를 의미한다는 것을 깨달았습니다.
```

devbattery reviewed (commented):
영한님 강의를 보고 진행했다 보니, 영한님이 id로 만드신 건 말 그대로 index였습니다. (Long)
이 미션에서 원하는 것은 String 형태의 아이디(userId)기 때문에 변경했습니다.
**(Long 형의 id) -> (String 형의 userId)**

아직 수요일인데도 불구하고 벌써 리뷰 남겨주셔서 정말 감사드립니다! 열심히 하겠습니다 🙇🏻

Commit:
fix: get 방식에서 post 방식으로 변경

Commit:
feat: UserRepository 객체를 관리하는 Service 생성

UserRepository 객체를 생성자로 받아와서 관리한다.
userId가 중복되는지 판별하고, 저장해주는 회원가입 기능을 갖고 있다.
user 전체를 찾는 기능도 갖고 있다.

Commit:
feat: 회원가입 할 때 입력받은 값들을 받아올 수 있도록 생성자 생성

model에 저장한 값을 User 클래스 안에 넣어줄 예정이다.

Commit:
feat: UserService 객체를 관리하도록 생성자 추가

Commit:
feat: 사용자 추가 기능 구현

전에 User와 UserForm을 수정한 이유가 바로 이것이다. 사용자를 받아서 users에 리다이렉트 되도록 한다.

Commit:
feat: 리다이렉트된 /users를 list.html에 리턴되도록 구현

이게 가능한 이유는 상대 경로로 넘겨줄 때 ViewResolver가 경로를 추가해주기 때문

Commit:
feat: model로 받아온 리스트를 각각 테이블에 전달

하지만 userId, name, email에 노란색 밑줄로 Cannot resolve가 뜨는데 정상적으로 넘어옴. 이 이유는 모르겠음

Commit:
fix: HTML 코드 중복 분리 수정

form에서만 header부분의 css 경로가 살짝 달랐고, navbar-fixed-top.html에서 빼먹었던 부분을 추가했다.

Commit:
feat: profile.html로 userId의 user 객체를 넘겨주는 메서드 추가

디버깅으로 어쩌구/users/입력한ID 입력하니 잘 넘어오는 것을 확인

Commit:
fix: 경로 수정

Commit:
fix: header 분리

header와 header2로 분리하여 각 파일에 맞는 frag로 정의했다

Commit:
fix: header2 경로 수정

예전과 똑같은 실수함. ../css로 변경

Commit:
frag: index와 index가 아닌 html 분리

index에서 접근할 때는 "/user/어쩌구"가 맞지만
거기로 이동한 후에는 "/어쩌구"가 맞다.
이렇게 하면 로그인으로 이동한 후에 계속 로그인을 클릭해도 에러가 뜨지 않는다.

Commit:
fix: redirect 이해 후 적용

redirect를 잘 이해하지 못 하고 있어서 Direct 메서드에서 그냥 /login에 가게 만드는 실수를 저질렀다.
직접 몇번 실험해 보면서 이럴 때 redirect를 쓰는 것임을 깨달았다.

Commit:
fix: 절대 경로로 수정

상대 경로로 했기 때문에 header2, for-index와 같은 파일들이 추가적으로 생성할 수밖에 없었다. 절대 경로로 수정한 이후로는 한 종류의 html 파일로만 적용이 가능하게 됐다

Commit:
fix: Direct 메서드 삭제

html에서 상대 경로로 전달해줬었기 때문에 상대경로/상대경로 이런 식으로 계속 경로 추가가 됐다. 하지만 절대 경로로 바꾸면 이 모든 것이 해결된다.

Commit:
feat: 회원가입 후 프로필로 넘어가는 기능 구현

회원가입 했을 때의 userId를 바탕으로 하여 경로에 추가해준다. 그러면 그 userID의 profile로 연결이 된다.

Commit:
feat: userId와 email 연결

profile.html에 model이 넘겨줬으니 그 정보를 바탕으로 출력

Commit:
fix: header2, for-index 없앤 결과 적용

header2를 없애고, for-index를 없앴으니 그것을 user 폴더 내의 파일과 index에도 적용시켜야 함

Commit:
fix: save 메서드의 반환 타입을 void로 변경

return 값이 존재할 이유가 아직까지는 없다.

Commit:
feat: 사용자 아이디를 클릭하면 그 프로필로 이동하게끔 구현

Posts의 각 게시물마다 작성자를 클릭하면 이처럼 되도록 구현 예정

Commit:
frag: qna 폴더 파일 templates로 이동 후 중복 분리

Commit:
feat: 각 파일에 qna 연결 후 경로 매핑

WebMVCConfigurer를 사용해서 매핑 방법 바꿔보는 것도 생각하기

Commit:
feat: 글쓴이, 제목, 내용 정보를 담고 있는 Article 클래스 생성

Commit:
feat: 글쓴이, 제목, 내용 정보를 처리하는 dto 생성

Commit:
feat: Article 정보를 저장하고, 찾는 인터페이스 생성

Commit:
feat: ArticleRepository를 구현하는 클래스 생성

Commit:
feat: ArticleRepository를 객체로 갖고 관리하는 서비스 생성

Commit:
style: 변수 이름 변경

write와 contents로 통일해주었다. html 파일의 id명을 참고했다.

Commit:
style: Form의 변수 이름 변경

write와 contents로 통일해주었다. html 파일의 id명을 참고했다.

Commit:
feat: 글 쓴 정보를 메인 페이지로 리다이렉트 하도록 구현

MvcConfig를 사용하려 했지만 잘 되지 않아 TODO로 남겨두었다.
글 썼던 정보는 form으로 다 잘 넘어온다.

Commit:
feat: WebMVCConfigurer를 적용하여 추가

WebMvcConfigurerAdapter가 deprecated 되었으므로 WebMvcConfigurer로 사용

Commit:
fix: "/"로 GetMapping 받아서 ArticleController에서 오류 났던 것 수정

Commit:
feat: index.html로 List 보내는 메서드 작성

Commit:
feat: /questions/post를 PostMapping 하기 위해 추가

Commit:
feat: index에 게시물이 계속 나열되도록 기능 추가

클릭하면 articles/아이디의 링크로 이동하므로 이에 대한 수정이 필요함

Commit:
fix: 게시물을 클릭하면 index 경로로 들어가게끔 설정

게시물을 추가할수록 index 1,2,3,4 이런 식으로 1씩 증가한다

Commit:
feat: Long형 변수 articleIndex를 추가

레포지토리, 서비스의 관계에 articleIndex가 엮일 수 있도록 했다.
게시물이 실행되면 articleIndex가 1씩 추가되도록 하였다.

Commit:
feat: show.html에 articleIndex를 넘겨주도록 구현

이를 위해 findByArticleIndex 메서드를 만들었음

Commit:
feat: 게시물에 관련된 정보를 표시할 수 있도록 타임리프 문법 추가

Commit:
feat: subnav에 질문하기 버튼 추가

active 타입이 고정되어 있어서 해결하려 했지만 현재 내 css 실력으로는 불가능

Commit:
feat: 수정 정보를 입력하는 폼 생성

form.html를 복사하여 수정하였다.

Commit:
feat: 수정 버튼을 누르면 userId의 경로로 가게끔 추가

정확히는 users/update/{userId}의 경로로 가게 했다.

Commit:
feat: 수정 버튼을 누르면 그 userId에 맞는 경로로 이동하도록 구현

Commit:
fix: label 추가

Commit:
style: button 네이밍 변경

회원가입에서 수정으로 변경

Commit:
feat: MvcConfig를 컨트롤러에 적용

MvcConfig는 WebMvcConfigurer를 구현한 Configuration 애노테이션을 사용하는 클래스

Commit:
bug: 회원 정보 수정하는 기능 추가 후 버그

수정한다는 개념이 제대로 머리에 박히질 않는다. 이론 공부 후 다시 구현해야겠다.

kyupid commented on src/main/java/kr/codesqaud/cafe/config/MvcConfig.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public void addViewControllers(ViewControllerRegistry registry) {
+        registry.setOrder(Ordered.HIGHEST_PRECEDENCE);
```
이건 무슨 코드예요? registry 에 매핑된 url 들을 가장 먼저 사용하겠다는 뜻인가요?

devbattery replied:
루카스 미션 내에서 제공해준 코드입니다.
저도 어떤 기능인지 잘 모르겠어서 그룹리뷰와 스쿼드세션 때 이 코드에 대해 많은 얘기를 했습니다.
이 코드가 필요 없다고 하신 분도 계셨지만 직접 제가 실험해 보니 우선순위를 설정하지 않았을 때 페이지가 정상적으로 응답하지 않았습니다.

이 코드는 `registry`에서 뷰 컨트롤러의 우선순위를 설정합니다.
`HIGHEST_PRECEDENCE`가 결국 가장 높은 우선순위를 나타내는 것이니
다른 뷰 컨트롤러보다 `Ordered.HIGHEST_PRECEDENCE`가 먼저 호출된다고 이해했습니다.

chatgpt에서는 "다른 ViewController에서 처리할 수 없는 요청을 처리하거나, 먼저 처리해야 하는 공통 기능을 구현하는 등의 용도로 사용될 수 있다"라고 하는데 현재 지식으로는 이해하기 힘들어 더 공부해나가야 할 것 같습니다.

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,45 @@
+    @PostMapping("/questions/post")
+    public String postArticle(ArticleForm form) {
+        Article article = new Article(form.getWriter(), form.getTitle(), form.getContents());
+        article.setWriter(article.getWriter());
```
writer 는 이미 생성자를 통해 값이 할당되었습니다. 따로 이유가 있을까요?

devbattery replied:
감사합니다. 이미 할당되어서 25번째 줄이 필요 없는 게 맞는 것 같습니다.
지워도 잘 작동됩니다 😄

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+    @PostMapping("/user/create")
+    public String create(UserForm form) {
+        User user = new User(form.getUserId(), form.getPassword(), form.getName(), form.getEmail());
+        user.setName(form.getName());
```
생성자를 통해 이미 할당 받은 name 을 다시 set 하는 이유가 있을까요?

devbattery replied:
위에서 말씀해주신 것처럼 이 부분도 없애야 하는 게 맞는 것 같습니다.
지우고 정상적으로 실행되었습니다!

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+        List<User> users = userService.findUsers();
+        model.addAttribute("users", users);
+        return "user/list";
+        // gradle -> 타임리프 ViewResolver가 이걸 해줌
+        // "templates/" + "user/list" + ".html"
```
코드를 아무리 나름대로 깔끔하게 작성해도 다른 사람이 봤을 때 그 의미가 분명히 드러나지 않는 곳이 있습니다. 그래서 주석을 남기는 것은 좋은 습관이라고 생각해요. 

여기에서는 개인 학습 목적용으로 이렇게 남겨 놓으신거죠? 이것 외에도 뭔가 다음에 보았을 때 기억에 잘 안남을 것 같은거는 적당히 주석을 잘 달아놓으시면 공부하는 데 더 도움이 될거예요 👍😊

devbattery replied:
감사합니다! 😭 제가 코딩테스트 연습할 때는 주석을 학습 목적으로 주렁주렁 달아 놓는데
거기에서 하는 만큼의 반이라도 가져와서 앞으로는 더 많은 (필요한) 주석을 잘 남기도록 하겠습니다

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+
+    @GetMapping("/users/{userId}")
+    public String findProfile(@PathVariable("userId") String userId, Model model) {
+        User user = userService.findByUserId(userId).get();
```
Optional 에서 get() 하는 것은 무엇을 의미하는지 생각해볼까요? 아마 인텔리제이에서 해당 코드에 마우스 커서 가져다대면 힌트가 있을 겁니다.

devbattery replied:
`isPresent()`에서 체크가 필요하다는 힌트가 있네요.!

```java
        User user;
        if (userService.findByUserId(userId).isPresent()) {
            user = userService.findByUserId(userId).get();
        } else {
            throw new NoSuchElementException();
        }

        model.addAttribute("user", user);
```

위와 같이 수정해 보았습니다.
맞게 수정한 건지는 확실하질 않지만 더 이상 인텔리제이에서 노란줄이 뜨진 않고 프로필도 정상적으로 출력됩니다.

kyupid replied:
넵 인텔리제이에서 기본으로 제공 해주는 기능이나 SonarLint, Copilot 같은 보조 도구들을 잘 활용하시면 학습하시는 데에 도움이 많이 될거예요.

kyupid replied:
Optional 사용은 https://mangkyu.tistory.com/203 이 문서를 참고하시면 좀 더 나은 코드를 작성하는 데에 도움이 될 것 같습니다.

devbattery replied:
감사합니다! 제가 안 좋은 것들을 다 하고 있었네요 🥲

kyupid replied:
아녜요 설리도 충분히 잘 하셨어요. 잘못하고 있다고 지적하는 게 아닙니다. 더 잘하실 수 있게 공부자료를 찾아 드린거뿐입니다 😅

devbattery replied:
아이고 너무 감사드립니다 😭
막 지적해주셔도 괜찮은데 배려까지 해주시네요 제공해주시는 자료 꼭꼭 씹어먹겠습니다!

kyupid commented on src/main/java/kr/codesqaud/cafe/domain/Article.java:
Article 포함 및 다른 클래스에서 사용하지 않는 메소드가 많아 보입니다. 삭제해주는 게 좋을 것 같아요. 

duplicated comment: https://github.com/codesquad-members-2023/be-java-cafe-max/pull/4/#discussion_r1152038175

devbattery replied:
분명 Article과 ArticleForm의 setter 메서드는 쓰이지 않는다고 되어 있는데
특정 클래스에서 이 메서드들을 삭제하면 글쓰기 행위를 한 후 제대로된 출력이 되질 않습니다.

제가 지운 메서드는 총 3종류입니다.
- `setWriter()`
- `setTitle()`
- `setContents()`

(dto) ArticleForm에서 위 메서드들을 삭제했을 때는 제대로 출력이 되질 않는 오류가 발생합니다.
하지만 (domain) Article에서 위 메서드들을 삭제했을 때는 정상적으로 동작이 됩니다.

결국 컨트롤러와 직접적인 접근이 이루어지는 것이 dto이기 때문인가?라는 생각이 들어
dto에 대한 정확한 동작 원리를 모르는 현재로서는 더 공부하며 깨달아 보겠습니다 감사합니다!

kyupid replied:
엇 그렇군요 저도 다시 보도록 하겠습니다. 제가 잘못 본거 같네요

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public void save(Article article) {
+        article.setArticleIndex(++sequence);
+        articleMap.put(article.getArticleIndex(), article);
```
```suggestion
        articleMap.put(++sequence, article);
```
위와 같이하는 게 더 좋을 것 같습니다.

devbattery replied:
아하 감사합니다 !!
그렇다면 Article 클래스에서 Long형 articleIndex라는 필드를 갖고 있지 않는 게 더 좋을 것 같다는 말씀이실까요??

kyupid replied:
코드 한줄을 줄일 수 있어서 그렇게 작성해봤어요. 여기에서 index 를 삭제해라는 의미는 없었습니다

devbattery replied:
엇 그러면 article에 있는 articleIndex를 Set하고 Get 과정이 없어지게 된다는 것인데
이렇게 되면 articleIndex가 있는 의미가 없어져버리거든요!
그 의미에서 말씀드렸습니다 😲

kyupid replied:
아하 이해했습니다. 제가 착각했어요.

추가 의견이 있는데 저게 Article 에 대한 식별값이면 Index 라는 네이밍보다 Id 가 더 적합하지 않을까요?

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,32 @@
+        articleMap.put(article.getArticleIndex(), article);
+    }
+
+    @Override
+    public List<Article> findAll() {
+        return new ArrayList<>(articleMap.values());
+    }
```
그냥 articleMap 을 반환하면 안될까요? 꼭 새로운 List 로 반환해야하나요? 어떤 이유로 이렇게 작성하신지 궁금해요.

devbattery replied:
결국 이 메서드가 ArticleController의 `articleList()` 메서드로 호출이 되는데,
이때 아래와 같은 형식으로 List형 articles로 저장이 됩니다.

model에 추가해줄 때 List가 아니라 Map이라면 문제가 발생하지 않을까 하여
List로 작성하는 것을 자연스럽게 받아들인 것 같습니다 !

```java
List<Article> articles = articleService.findArticles();
model.addAttribute("articles", articles);
```

kyupid replied:
앞단에서 처리는 어떻게 해주는지 잘 모르겠습니다. 백엔드만 생각해볼게요. 제 생각에 만약에 MemoryRepository 내에 있는 캐시되어 있는 map 을 그대로 반환하면 동시성 문제가 있지 않을까 싶어요. 그래서 작성하신 것처럼 새 컬렉션 객체에 담아서 리턴하는 게 올바른 방법인거 같구요. 혹시 방어적 복사를 모르신다면 한번 찾아보시면 좋을 것 같아요!

devbattery replied:
아 이렇게 방어적 복사로 하면 변경에 대해서는 원본과 복사본이 서로 영향을 받지 않게 되네요..!!
말씀 안 해주셨으면 정확한 원리를 모르고 넘어갈 뻔했는데 감사합니다 ㅠㅠㅠ

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void save(User user) {
+        if (userMap.containsKey(user.getUserId())) {
+            return;
+        }
```
join 할때 이미 존재하는 user 는 어떻게 처리되어질까요?

devbattery replied:
앗 그냥 단순하게 반환만 되네요..!

이럴 때는 Valid로 무언가를 해주는 게 더 좋을 것 같은데
지금 상황에서는 return보단는 예외를 던지는 것이 더 바람직 할 것 같습니다.
혹시 이 접근이 맞을까요..?

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Override
+    public Optional<User> findByUserId(String userId) {
+        return Optional.ofNullable(userMap.get(userId));
+    }
```
위에서 말씀드렸듯이 Optional.get() 이 어떤 의미를 가지는지 한번 생각해보세요.

devbattery replied:
감사합니다! 위에서 한 것처럼 맞는 방법인지는 모르겠지만 `update()` 메서드를 수정하겠습니다!

```java
        User findUser;
        if (findByUserId(userId).isPresent()) {
            findUser = findByUserId(userId).get();
        } else {
            throw new NoSuchElementException();
        }
```

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,38 @@
+        return new ArrayList<>(userMap.values());
+    }
+
+    @Override
+    public void update(String userId, User updateUser) {
+        User findUser = findByUserId(userId).get();
+        findUser.setName(updateUser.getName());
+        findUser.setEmail(updateUser.getEmail());
+        findUser.setPassword(updateUser.getPassword());
+    }
```
User 를 update 할 때 항상 name, email, password 를 같이 업데이트하도록 구현하셨네요. **하나의 기능**으로써 동작하는 것 같은데 좀 더 개선해볼 수 없을까요? 😊

devbattery replied:
앗 정말 그렇네요..!
제가 지금 떠오르는 방법으로는 이 update 메서드를 updateName, updateEmail, updatePassword로 세분화하여
각각 총 3개의 메서드로 만들어서 다른 하나의 메서드로 관리하는 방법이 떠오르는데 kyu께서는 어떻게 생각하시나요?

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/UserRepository.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface UserRepository {
+    void save(User user);
+
+    Optional<User> findByUserId(String userId);
+
+    List<User> findAll();
+
+    void update(String userId, User updateUser);
+}
```
꼭 인터페이스를 하나 두어야하나요? 현재 MemoryXXXRepository 와 같이 구현체가 하나 밖에 없는데 인터페이스가 있어야하나요? 술리는 어떻게 생각하세요?

devbattery replied:
사실 이 부분은 영한님의 강의에서 사용하셔서 그대로 사용한 것이 없지 않아 있습니다.
말씀하신 것처럼 인터페이스를 사용하는 의미가 제 코드에서는 퇴색된 것 같습니다.

UserRepository와 ArticleRepsitory의 각각 인터페이스에서 겹치는 메서드도
`save()`와 `findAll()`밖에 없으니 제네릭으로 선언해도 별 믜미가 없을 것 같습니다.

하지만 제가 xxxxService의 서비스에서 구현 객체(MemoryUesrRepository)가 아닌, UserRepository를 갖고 있으니
추상화가 되므로, 변경/유지 면에서는 좋다고 생각했습니다.

앞으로 이런 상황에서는 초반부터 만들어나가는 것에 대해 좀 더 회의적으로 바라보며 접근하도록 하겠습니다 감사합니다!

kyupid replied:
한번 생각해보시라는 차원에서 던진 질문이었어요. 지금처럼 나름의 대답을 가지고 있으시면 됩니다 👍

kyupid commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -0,0 +1,28 @@
+
+@Service
+public class ArticleService {
+    private final ArticleRepository articleRepository;
+
+    public ArticleService(ArticleRepository articleRepository) {
+        this.articleRepository = articleRepository;
+    }
```
```suggestion
@Autowired
ArticleRepository articleRepository;
```
이렇게 사용하면 코드가 더 깔끔하지 않을까요? 왜 생성자 주입을 사용하셨나요?

devbattery replied:
생성자가 하나 있을 때는 `@Autowired`를 사용하는 것이 의미 없다는 말을 들었습니다.
그 말을 듣기 전, 이를 시도해 보았었는데 `Field injection is not recommended`라는 것을 인텔리제이가 알려주었습니다.

사실 여기까지만 알고 그냥 넘어갔었지만 지금 kyu의 말씀으로 의존성 주입에 대해 더 공부해볼 것 같습니다.
더 생각해볼 기회를 만들어주셔서 정말 감사드립니다 😄

kyupid commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,41 @@
+        userRepository.save(user);
+    }
+
+    private void validateDuplicateUser(User user) {
+        userRepository.findByUserId(user.getUserId())
+                .ifPresent(u -> {
+                    throw new IllegalStateException("이미 존재하는 회원입니다.");
+                });
+    }
```
두 가지 의견있습니다.

1.  밑에 서비스에서 만드신 findByUserId 와 똑같이 userRepository.findByUserId 가 겹치는 부분이 있는데 개선 가능할까요?
2. IllegalStateException 은 메소드 호출에는 부적합한 상태일때 던지는 예외인데  "이미 존재하는 회원" 이라는 게 validate 하는 메소드를 호출하는데 부적합하다고는 생각이 안듭니다. 제 개인 의견이긴 하지만 다음엔 다른 Exception 이나 커스텀하는 것도 고려해보시면 좋을 것 같습니다

devbattery replied:
1. 어엇 그렇네요!! 결국 최종적으로 `findByUserId()`가 실행되는 것이 이 서비스이기 때문에 아래와 같이 처리하는 게 맞다고 생각하여 수정해 보았습니다.

```java
    private void validateDuplicateUser(User user) {
        findByUserId(user.getUserId())
                .ifPresent(u -> {
                    throw new IllegalStateException("이미 존재하는 회원입니다.");
                });
    }
```

2. Exception 자체도 홈페이지를 사용하는 유저의 입장에서는 보이지 않기 때문에 부적절하다고 생각이 들었는데 Exception의 타입도 잘못되었네요 🥲.. 그리고 다른 동료들의 코드에서 `@Valid`로 더 멋지게 처리하는 방법이 있던데 저도 앞으로는 이 애노테이션을 사용하여 예외처리를 할 수 있또록 노력해보겠습니다!

kyupid replied:
멋지게 처리하는 게 더 나은 방법은 아니라고 생각합니다만 ㅎㅎ 그런 다른 방법들도 다 배워놓고 선택할 수 있는 정도의 능력을 기르는 건 좋겠죠.

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+
+    @PostMapping("/user/create")
+    public String create(UserForm form) {
+        User user = new User(form.getUserId(), form.getPassword(), form.getName(), form.getEmail());
```
지금은 괜찮지만 User 에 더 많은 필드가 있으면 생성자 파라미터가 굉장히 많고 헷갈릴거 같은데요. 다음엔 빌더 패턴도 한 번 고려해봅시다. 😊 

[참고 자료](https://johngrib.github.io/wiki/pattern/builder/#effective-java%EC%9D%98-%EB%B9%8C%EB%8D%94-%ED%8C%A8%ED%84%B4)

devbattery replied:
> - 규칙 2. 생성자 인자가 많을 때는 Builder 패턴 적용을 고려하라 (2판)
> - 아이템 2. 생성자에 매개변수가 많다면 빌더를 고려하라 (3판)

꼭 기억해서 다음부터 적용할 수 있도록 하겠습니다 😄

혹시 ArticleForm형식의 form만을 넘겨서 Article 클래스 내에서 처리하는 방법은 좋지 않을까요?

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+        return "user/updateForm";
+    }
+
+    @PostMapping("/users/update/{userId}")
```
따로 URL을 어떤식으로 매핑할 것인지 전략을 생각하시고 작성하신 부분인가요? 만약에 그렇다면 어떤 식으로 생각해보셨는지 궁금합니다 😊

devbattery replied:
어제 수정한 부분이 있어 `feat: put 방식 추가 후 타임리프 링크 설정`이라는 메시지명으로 다시 커밋하였습니다.


```java
    @PutMapping("/users/update/{userId}")
    public String putUpdate(@PathVariable String userId, User user) {
        userService.updateUser(userId, user);
        return "redirect:/users";
    }
```

```html
<form name="question" method="post" action="#" th:action="@{/users/update/{userId}(userId=${userUpdated.userId})}">
                <input type="hidden" name="_method" value="PUT" />
```

PutMapping 형식으로  수정하였습니다.
`updateform.html`의 "수정" 버튼을 누르면 `userService.updateUser(userId, user);`가 실행이 되어 업데이트가 되고
`/users`로 리다이렉트 되도록 할 생각이었습니다.
막상 실행을 시켜 보니 제가 의도한 바와는 다르게 흘러가며 혼란스러웠습니다.

이런 경험을 하고 나니 아직 제대로 된 원리를 이해하지 못한다는 생각이 들었습니다.
이때부터 선택미션 구현에 더 손을 대지 않고, 지금까지 구현했던 것에 대한 이론 강의를 듣게 된 것 같습니다.

kyupid replied:
제 의견이 잘 전달이 안된것 같아요. 수정 요청을 한 것이 아니라 단순히 URL 을 어딴 규칙으로 만들려고 하셨는지가 궁금했고 있었다면 설명을 부탁드리려고 했던거였습니다!

devbattery replied:
아하! 그렇군요 😲 죄송함니닷

회원 리스트는 users로 리다이렉트 되도록해서 localhost:8080/users에서 회원 정보 수정을 누를 수 있습니다.
거기에서 수정을 누르면 localhost:8080/users/update/유저아이디 이 경로로 가게끔 하였습니다.

update라는 경로를 추가한 것은 수정하는 페이지를 임의적으로 나타낸 것이고,
{userId}는 그 유저 아이디를 받아오게끔 하였습니다 :)

kyupid reviewed (approved):
안녕하세요 술리🕺 첫번째 PR 학습하신다고 고생많으셨습니다.

저도 처음에 코드스쿼드에서 아무것도 몰라 난관이 많아서 작성하신 고민과 해결 부분 보니 공감이 많이 되네요. 계속 전진하시면 분명히 좋은 결과가 있을거예요. 😊

동시성 문제가 일어나지 않을거 같다고 HashMap 으로 본인의 생각을 담아 코드를 작성하신 것은 정말 훌륭하신 것 같습니다. 저도 그렇고 많은 분들이 처음에 따라하기에도 바빠서 그런 생각을 잘 못하거든요. 개발자는 정말 코드 한줄한줄 본인이 왜 그렇게 작성한 것인지 틀려도 이유가 있어야한다고 생각합니다. 잘하셨어요 👏

밑에 질문에 대해서 답변 드리겠습니다.
```
Map<String, User> vs Map<Long, User>
MemoryArticleRepository 클래스에서는 글 목록의 index를 카운트해줘야 하므로 Long형으로 저항 없이 썼지만
MemoryUserRepository 클래스에서는 user의 index가 필요 없으므로 String형으로 사용했습니다.
이 주제에 대하여 동료들과 토론을 했지만 결론은 어떤 상황에서도 Long형으로 정의하는 것이 맞다는 것이었습니다.
kyu께서는 이에 대해 어떻게 생각하시는지 여쭤보고 싶습니다. 어떤 상황에서든 index를 추가하여 Long형으로 정의하는 것이 맞을까요?
```
"어떤 상황에서든" 이라고 단언하시어 생각하시면 절대 안됩니다. 개발하면서 생각하시는 것보다 무궁무진하게 다양한 상황들이 존재해요. 

일단 답변을 드리면 MemoryUserRepository 이니까 메모리에서 String 과 Long 으로 접근했을 때 어떤 성능적인 아웃풋을 내는지 생각해볼거 같애요. **그런데 가장 중요한 것은 이게 정말 의미있는 Issue 인지부터 생각해볼 것 같습니다.** 메모리에서 사용한다면 단순 테스트용일 것 같은데 과연 이것이 Issue 일까요?

Commit:
fix: Optional 에서 get() 하는 것은 isPresent() 체크 필요

Commit:
fix: 불필요한 코드 제거

set을 한 번 더 해줄 필요가 없었음

Commit:
feat: put 방식 추가 후 타임리프 링크 설정

Commit:
delete: Article 클래스의 불필요한 세터 삭제

devbattery reviewed (commented):
이번주 정말 저보다도 훨씬 더 고생 많으셨습니다!! 감사합니다 😭
저도 언젠가 kyu처럼 실력 좋은 개발자가 될 수 있도록 꾸준히 노력해보겠습니다!!

답변해주신 코멘트를 보니 확실히 제가 평소에 잘못 생각하고 있던 부분이 무엇인지 깨달았습니다.
개발에 있어서 절대 "어떤 상황에서든"이 있으면 좋지 않고, 상황에 따라 언제든지 바뀔 수 있다는 것을 인지하겠습니다 😤

말씀하신 것처럼 이 부분은 결국 DB로 대체될 것이기도 하고, 메모리에서 사용되는 것인데 Issue라고 할 만큼의 의미는 없었던 것 같습니다. 결국 이 부분이 이슈가 된다고 해도 앞으로 **성능**적인 부분에 맞춰서 생각해보도록 하겠습니다

너무너무 감사드립니다 🙇🏻

kyupid commented:
@won4885 다음부터는 의도를 명확하게 하기 위해서 이해하시기 편하게 코멘트에 따로 라벨을 달아놓을게요. 관련해서 다음 PR 때 다시 언급드리겠습니다.

그리고 전체적으로 질문이 너무 추상적이었나요? 저도 리뷰어는 처음이라 부족한 점이 많을 수도 있습니다. 제 리뷰에 대한 피드백을 주시면 적극 의견 수용하겠습니다. 어떻게 하면 더 도움을 드릴 수 있을까요?

devbattery commented:
@kyupid 

> 다음부터는 의도를 명확하게 하기 위해서 이해하시기 편하게 코멘트에 따로 라벨을 달아놓을게요. 관련해서 다음 PR 때 다시 언급드리겠습니다.

넵 알겠습니다!

> 그리고 전체적으로 질문이 너무 추상적이었나요? 저도 리뷰어는 처음이라 부족한 점이 많을 수도 있습니다. 제 리뷰에 대한 피드백을 주시면 적극 의견 수용하겠습니다. 어떻게 하면 더 도움을 드릴 수 있을까요?

첫 주라 맞춰가는 과정일 뿐이라서 이번주는 어쩔 수 없었다고 생각해요! 😁
kyu의 질문이 정말로 궁금해서 하신 질문인지, 아니면 제 학습을 위한 질문인지
이제는 파악이 가능해져서 이번주부터는 저만 부족한 점을 채워나가면 될 것 같아요!

오히려 제가 이번주부터는 제 학습 목적으로도 그렇고, 코드를 남이 알아 보기 쉽도록 하는 것도 그렇고
주석 다는 연습을 하면서 더 이해하기 쉬우시도록 노력해 보겠습니다!

리뷰 열심히 달아주셔서 진심으로 너무 감사드립니다 🙇🏻

kyupid commented:
@won4885 더 궁금하신 점은 다음 PR 에 다시 언급해주시면 답변드리겠습니다. 고생하셨습니다. 이 PR 은 마무리하시죠 😊
